client/driver_manager_test.go

```diff
@@ -0,0 +1,186 @@
+	"time"
+
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/client/pluginmanager/drivermanager"
+	"github.com/hashicorp/nomad/helper/pluginutils/catalog"
+	nconfig "github.com/hashicorp/nomad/nomad/structs/config"
+	"github.com/hashicorp/nomad/testutil"
+	"github.com/stretchr/testify/require"
+)
+
+// TestDriverManager_Fingerprint_Run asserts that node is populated with
+// driver fingerprints
+func TestDriverManager_Fingerprint_Run(t *testing.T) {
+	t.Parallel()
+
+	testClient, cleanup := TestClient(t, nil)
+	defer cleanup()
+
+	dm := drivermanager.New(&drivermanager.Config{
+		Logger:              testClient.logger,
+		Loader:              testClient.config.PluginSingletonLoader,
+		PluginConfig:        testClient.configCopy.NomadPluginConfig(),
+		Updater:             testClient.updateNodeFromDriver,
+		EventHandlerFactory: testClient.GetTaskEventHandler,
+		State:               testClient.stateDB,
+	})
+
+	go dm.Run()
+	defer dm.Shutdown()
+
+	testutil.WaitForResult(func() (bool, error) {
+		node := testClient.configCopy.Node
+
+		d, ok := node.Drivers["mock_driver"]
+		if !ok {
+			return false, fmt.Errorf("mock_driver driver is not present: %+v", node.Drivers)
+		}
+
+		if !d.Detected || !d.Healthy {
+			return false, fmt.Errorf("mock_driver driver is not marked healthy: %+v", d)
+		}
+
+		return true, nil
+	}, func(err error) {
+		require.NoError(t, err)
+	})
+}
+
+// TestDriverManager_Fingerprint_Run asserts that node is populated with
+// driver fingerprints and it's updated periodically
+func TestDriverManager_Fingerprint_Periodic(t *testing.T) {
+	t.Parallel()
+
+	testClient, cleanup := TestClient(t, func(c *config.Config) {
+		pluginConfig := []*nconfig.PluginConfig{
+			{
+				Name: "mock_driver",
+				Config: map[string]interface{}{
+					"shutdown_periodic_after":    true,
+					"shutdown_periodic_duration": 2 * time.Second,
+				},
+			},
+		}
+
+		c.PluginLoader = catalog.TestPluginLoaderWithOptions(t, "", map[string]string{}, pluginConfig)
+
+	})
+	defer cleanup()
+
+	dm := drivermanager.New(&drivermanager.Config{
+		Logger:              testClient.logger,
+		Loader:              testClient.config.PluginSingletonLoader,
+		PluginConfig:        testClient.configCopy.NomadPluginConfig(),
+		Updater:             testClient.updateNodeFromDriver,
+		EventHandlerFactory: testClient.GetTaskEventHandler,
+		State:               testClient.stateDB,
+	})
+
+	go dm.Run()
+	defer dm.Shutdown()
+
+	// we get a healthy mock_driver first
+	testutil.WaitForResult(func() (bool, error) {
+		node := testClient.configCopy.Node
+
+		d, ok := node.Drivers["mock_driver"]
+		if !ok {
+			return false, fmt.Errorf("mock_driver driver is not present: %+v", node.Drivers)
+		}
+
+		if !d.Detected || !d.Healthy {
+			return false, fmt.Errorf("mock_driver driver is not marked healthy: %+v", d)
+		}
+
+		return true, nil
+	}, func(err error) {
+		require.NoError(t, err)
+	})
+
+	// we eventually get an unhealthy one eventually
```

Comment:
two eventually

Reply:
I think we only check if we get a single unhealthy fingerprint